ocaml/libs/lru-cache/lru.ml

```diff
@@ -0,0 +1,294 @@
+    let rec loop acc = function
+      | None ->
+          acc
+      | Some n ->
+          loop (f n.value acc) n.prev
+    in
+    loop zero t.last
+
+  let to_list t = foldr (fun x xs -> x :: xs) t []
+
+  let from_list xs =
+    let t = create () in
+    List.iter (fun x -> append t (node x)) xs ;
+    t
+end
+
+(** A store for key/value pairs of type ['k] and ['v]. The main store
+      is [table] that maps a key to a key/value node. Every item in the
+      store is also a member in the [queue]. The [queue] keeps track of
+      which elements are looked up most often. The first element in the
+      [queue] is the least used one. *)
+type ('k, 'v) t = {
+    table: ('k, ('k * 'v) LL.node) Hashtbl.t
+  ; queue: ('k * 'v) LL.t
+  ; cap: int  (** max capacity of table and queue *)
+  ; mutable entries: int  (** actual capacity of table and queue *)
+  ; lock: Mutex.t  (** lock while operating on this value *)
+}
+
+let locked m f =
+  let finally () = Mutex.unlock m in
+  Mutex.lock m ; Fun.protect ~finally f
+
+(* All primed functions below are not thread safe because they are
+   manipulating state; we will use a lock to protect against concurrent
+   update. However, we have to do that on an outer layer such that we
+   can use these functions internally after we obtained the lock. *)
+
+module Unsafe = struct
+  let create' capacity =
+    if capacity <= 0 then
+      invalid_arg "%s: capacity needs to be postive" __FUNCTION__ ;
+    {
+      table= Hashtbl.create capacity
+    ; queue= LL.create ()
+    ; cap= capacity
+    ; entries= 0
+    ; lock= Mutex.create ()
+    }
+
+  let size' t = t.entries
+
+  let cap' t = t.cap
+
+  let to_list' t = LL.to_list t.queue
+
+  (** [lookup] an entry based on its [key]; this may fail or succeeed.
+      In the success case, the entry is moved to the tail of the
+      [queue]. Hnece, the least-used entry is at the front. *)
+  let lookup' t key =
+    match Hashtbl.find_opt t.table key with
+    | Some v ->
+        LL.drop t.queue v ;
+        LL.append t.queue v ;
+        Some (LL.value v |> snd)
+    | None ->
+        None
+
+  (** lookup and remove entry *)
+  let retrieve' t key =
+    match Hashtbl.find_opt t.table key with
+    | Some v ->
+        LL.drop t.queue v ;
+        Hashtbl.remove t.table key ;
+        t.entries <- t.entries - 1 ;
+        Some (LL.value v |> snd)
+    | None as result ->
+        result
+
+  (** [remove] an entry based on this [key] *)
+  let remove' t key = retrieve' t key |> ignore
+
+  (** [add] a new entry; do nothing if the entry exists. If the new
+      entry exceeds the capacity of the table, [true]
+      is returned and [false] otherwise. It signals the caller to [trim]
+      the table.*)
+  let add' t key value =
+    match lookup' t key with
+    | None ->
+        let node = LL.node (key, value) in
+        Hashtbl.add t.table key node ;
+        t.entries <- t.entries + 1 ;
+        LL.append t.queue node ;
+        t.entries > t.cap
```

Comment:
If we store anything in the LRU that requires handling before it can be removed, the client must be given an opportunity to do that. Hence I avoid removing elements when capacity is exceeded. I don't care so much about temporarily exceeding the capacity.